prog/hints.go

```diff
@@ -147,13 +146,45 @@ func checkDataArg(arg *DataArg, compMap CompMap, exec func()) {
 		original := make([]byte, 8)
 		copy(original, data[i:])
 		val := binary.LittleEndian.Uint64(original)
-		for _, replacer := range shrinkExpand(val, compMap, 64) {
+		for _, replacer := range shrinkExpand(val, compMap, 64, false) {
+			binary.LittleEndian.PutUint64(bytes, replacer)
+			copy(data[i:], bytes)
+			exec()
+		}
+		copy(data[i:], original)
+	}
+}
+
+func checkCompressedArg(arg *DataArg, compMap CompMap, exec func()) {
+	data0 := arg.Data()
+	if len(data0) == 0 {
+		return
+	}
+	imageMu.Lock()
```

Comment:
Did it work fine during long syzkaller runs?

I fear it might slow down everything -- hints can take quite a while if there are many, and it would just stall every proc that starts to mutate an fs image.

I wonder if it's worth it if we restructure the code to keep `exec()` out of the mutex. Maybe release the uncompressed memory before exec and, afterwards, decompress it again?

Reply:
Uploaded an optimized version that removes the mutex entirely (except for windows and 32-bit builds)